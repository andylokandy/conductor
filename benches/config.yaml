server:
  port: 9000

logger:
  level: error
  format: pretty

sources:
  - id: countries
    type: graphql
    config:
      endpoint: http://localhost:4000/graphql

endpoints:
  - path: /graphql
    from: countries

  - path: /test
    from: countries
