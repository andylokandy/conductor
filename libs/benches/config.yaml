server:
  port: 9000

logger:
  level: error
  format: pretty

sources:
  - id: countries
    type: graphql
    config:
      endpoint: http://localhost:4444/graphql

endpoints:
  - path: /graphql
    from: countries

  - path: /test
    from: countries
