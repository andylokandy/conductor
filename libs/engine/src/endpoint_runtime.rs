use conductor_config::EndpointDefinition;

#[derive(Debug)]
pub struct EndpointRuntime {
    pub config: EndpointDefinition,
}

impl EndpointRuntime {
    #[cfg(feature = "test_utils")]
    pub fn dummy() -> Self {
        EndpointRuntime {
            config: EndpointDefinition {
                from: "dummy".to_string(),
                path: "/".to_string(),
                plugins: None,
            },
        }
    }
}
