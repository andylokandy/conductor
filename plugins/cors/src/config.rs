use conductor_common::serde_utils::{
  JsonSchemaExample, JsonSchemaExampleMetadata, JsonSchemaExampleWrapperType,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// The `cors` plugin enables [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) configuration for your GraphQL API.
///
/// By using this plugin, you can define rules for allowing cross-origin requests to your GraphQL server. This is essential for web applications that need to interact with your API from different domains.
///
#[derive(Deserialize, Serialize, Debug, Clone, JsonSchema)]
#[schemars(example = "cors_plugin_example1")]
#[schemars(example = "cors_plugin_example2")]
#[schemars(example = "cors_plugin_example3")]
pub struct CorsPluginConfig {
  /// `Access-Control-Allow-Credentials`: Specifies whether to include credentials in the CORS headers. Credentials can include cookies, authorization headers, or TLS client certificates.
  /// Indicates whether the response to the request can be exposed when the credentials flag is true.
  #[serde(
    default = "default_boolean_false",
    skip_serializing_if = "Option::is_none"
  )]
  pub allow_credentials: Option<bool>,

  /// `Access-Control-Allow-Methods`: Defines the HTTP methods allowed when accessing the resource. This is used in response to a CORS preflight request.
  /// Specifies the method or methods allowed when accessing the resource in response to a preflight request.
  /// You can also specify a special value "*" to allow any HTTP method to access the resource.
  #[serde(default = "default_wildcard", skip_serializing_if = "Option::is_none")]
  pub allowed_methods: Option<String>,

  /// `Access-Control-Allow-Origin`: Determines which origins are allowed to access the resource. It can be a specific origin or a wildcard for allowing any origin.
  /// You can also specify a special value "*" to allow any origin to access the resource.
  /// You can also specify a special value "reflect" to allow the origin of the incoming request to access the resource.
  #[serde(default = "default_wildcard", skip_serializing_if = "Option::is_none")]
  pub allowed_origin: Option<String>,

  /// `Access-Control-Allow-Headers`: Lists the headers allowed in actual requests. This helps in specifying which headers can be used when making the actual request.
  /// Used in response to a preflight request to indicate which HTTP headers can be used when making the actual request.
  /// You can also specify a special value "*" to allow any headers to be used when making the actual request, and the `Access-Control-Request-Headers` will be used from the incoming request.
  #[serde(default = "default_wildcard", skip_serializing_if = "Option::is_none")]
  pub allowed_headers: Option<String>,

  /// `Access-Control-Expose-Headers`: The "Access-Control-Expose-Headers" response header allows a server to indicate which response headers should be made available to scripts running in the browser, in response to a cross-origin request.
  /// You can also specify a special value "*" to allow any headers to be exposed to scripts running in the browser.
  #[serde(default = "default_wildcard", skip_serializing_if = "Option::is_none")]
  pub exposed_headers: Option<String>,

  /// `Access-Control-Allow-Private-Network`: Indicates whether requests from private networks are allowed when originating from public networks.
  #[serde(
    default = "default_boolean_false",
    skip_serializing_if = "Option::is_none"
  )]
  pub allow_private_network: Option<bool>,

  /// `Access-Control-Max-Age`: Indicates how long the results of a preflight request can be cached.
  /// This field represents the duration in seconds.
  #[serde(default = "defualt_max_age", skip_serializing_if = "Option::is_none")]
  pub max_age: Option<u64>,
}

impl Default for CorsPluginConfig {
  fn default() -> Self {
    Self {
      allow_credentials: default_boolean_false(),
      allowed_methods: default_wildcard(),
      allowed_origin: default_wildcard(),
      allowed_headers: default_wildcard(),
      exposed_headers: default_wildcard(),
      allow_private_network: default_boolean_false(),
      max_age: defualt_max_age(),
    }
  }
}

fn default_wildcard() -> Option<String> {
  Some("*".to_string())
}

fn default_boolean_false() -> Option<bool> {
  Some(false)
}

fn defualt_max_age() -> Option<u64> {
  None
}

fn cors_plugin_example1() -> JsonSchemaExample<CorsPluginConfig> {
  JsonSchemaExample {
        metadata: JsonSchemaExampleMetadata::new(
            "Strict CORS",
            Some("This example demonstrates how to configure the CORS plugin with a strict list of methods, headers and origins."),
        ),
        wrapper: Some(JsonSchemaExampleWrapperType::Plugin { name: "cors".into() }),
        example: CorsPluginConfig {
                allow_credentials: Some(true),
                exposed_headers: None,
                allowed_methods: Some("GET, POST".into()),
                allowed_origin: Some("https://example.com".into()),
                allowed_headers: Some("Content-Type, Authorization".into()),
                allow_private_network: Some(false),
                max_age: Some(3600),
            },
    }
}

fn cors_plugin_example2() -> JsonSchemaExample<CorsPluginConfig> {
  JsonSchemaExample {
    metadata: JsonSchemaExampleMetadata::new(
      "Permissive CORS",
      Some("This example demonstrates how to configure the CORS plugin with a permissive setup."),
    ),
    wrapper: Some(JsonSchemaExampleWrapperType::Plugin {
      name: "cors".into(),
    }),
    example: CorsPluginConfig {
      allow_credentials: Some(true),
      exposed_headers: Some("*".into()),
      allowed_methods: Some("*".into()),
      allowed_origin: Some("*".into()),
      allowed_headers: Some("*".into()),
      allow_private_network: Some(true),
      max_age: Some(3600),
    },
  }
}

fn cors_plugin_example3() -> JsonSchemaExample<CorsPluginConfig> {
  JsonSchemaExample {
    metadata: JsonSchemaExampleMetadata::new(
      "Reflect Origin",
      Some(
        "This example demonstrates how to configure the CORS plugin with a reflect Origin setup.",
      ),
    ),
    wrapper: Some(JsonSchemaExampleWrapperType::Plugin {
      name: "cors".into(),
    }),
    example: CorsPluginConfig {
      allow_credentials: Some(true),
      exposed_headers: Some("*".into()),
      allowed_methods: Some("GET, POST".into()),
      allowed_origin: Some("reflect".into()),
      allowed_headers: Some("*".into()),
      allow_private_network: Some(false),
      max_age: Some(3600),
    },
  }
}
