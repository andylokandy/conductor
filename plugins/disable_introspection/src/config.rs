use conductor_common::{
  serde_utils::{JsonSchemaExample, JsonSchemaExampleMetadata, JsonSchemaExampleWrapperType},
  vrl_utils::VrlConfigReference,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Default, Deserialize, Serialize, Debug, Clone, JsonSchema)]
#[schemars(example = "disable_introspection_example1")]
#[schemars(example = "disable_introspection_example2")]
/// The `disable_introspection` plugin allows you to disable introspection for your GraphQL API.
///
/// A [GraphQL introspection query](https://graphql.org/learn/introspection/) is a special GraphQL query that returns information about the GraphQL schema of your API.
///
/// It it [recommended to disable introspection for production environments](https://escape.tech/blog/should-i-disable-introspection-in-graphql/), unless you have a specific use-case for it.
///
/// It can either disable introspection for all requests, or only for requests that match a specific condition (using VRL scripting language).
///
pub struct DisableIntrospectionPluginConfig {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  /// A VRL condition that determines whether to disable introspection for the request. This condition is evaluated only if the incoming GraphQL request is detected as an introspection query.
  ///
  /// The condition is evaluated in the context of the incoming request and have access to the metadata field `%downstream_http_req` (fields: `body`, `uri`, `query_string`, `method`, `headers`).
  ///
  /// The condition must return a boolean value: return `true` to continue and disable the introspection, and `false` to allow the introspection to run.
  ///
  /// In case of a runtime error, or an unexpected return value, the script will be ignored and introspection will be disabled for the incoming request.
  pub condition: Option<VrlConfigReference>,
}

fn disable_introspection_example1() -> JsonSchemaExample<DisableIntrospectionPluginConfig> {
  JsonSchemaExample {
    metadata: JsonSchemaExampleMetadata::new(
      "Disable Introspection",
      Some("This example disables introspection for all requests for the configured Endpoint."),
    ),
    wrapper: Some(JsonSchemaExampleWrapperType::Plugin {
      name: "disable_introspection".to_string(),
    }),
    example: DisableIntrospectionPluginConfig {
      ..Default::default()
    },
  }
}

fn disable_introspection_example2() -> JsonSchemaExample<DisableIntrospectionPluginConfig> {
  JsonSchemaExample {
        metadata: JsonSchemaExampleMetadata::new(
            "Conditional",
            Some(
                "This example disables introspection for all requests that doesn't have the \"bypass-introspection\" HTTP header.",
            ),
        ),
        wrapper: Some(JsonSchemaExampleWrapperType::Plugin {
            name: "disable_introspection".to_string(),
        }),
        example: DisableIntrospectionPluginConfig {
            condition: Some(VrlConfigReference::Inline { content: "%downstream_http_req.headers.\"bypass-introspection\" != \"1\"".to_string() }),
        },
    }
}
