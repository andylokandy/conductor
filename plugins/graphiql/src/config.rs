use conductor_common::serde_utils::{
  JsonSchemaExample, JsonSchemaExampleMetadata, JsonSchemaExampleWrapperType,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug, Clone, Default, JsonSchema)]
#[schemars(example = "graphiql_example")]
/// This plugin adds a GraphiQL interface to your Endpoint.
///
/// This plugin is rendering the GraphiQL interface for HTTP `GET` requests, that are not intercepted by other plugins.
pub struct GraphiQLPluginConfig {
  #[serde(
    default = "headers_editor_enabled_default_value",
    skip_serializing_if = "Option::is_none"
  )]
  /// Enable/disable the HTTP headers editor in the GraphiQL interface.
  pub headers_editor_enabled: Option<bool>,
}

fn graphiql_example() -> JsonSchemaExample<GraphiQLPluginConfig> {
  JsonSchemaExample {
    metadata: JsonSchemaExampleMetadata::new("Enable GraphiQL", None),
    wrapper: Some(JsonSchemaExampleWrapperType::Plugin {
      name: "graphiql".to_string(),
    }),
    example: GraphiQLPluginConfig {
      headers_editor_enabled: Default::default(),
    },
  }
}

fn headers_editor_enabled_default_value() -> Option<bool> {
  Some(true)
}

// At some point, it might be worth supporting more options. see:
// https://github.com/dotansimha/graphql-yoga/blob/main/packages/graphiql/src/YogaGraphiQL.tsx#L35
#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct GraphiQLSource {
  pub endpoint: String,
  pub query: String,
  #[serde(rename = "isHeadersEditorEnabled")]
  pub headers_editor_enabled: bool,
}
