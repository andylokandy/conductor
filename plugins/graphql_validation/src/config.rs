use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug, Clone, Default, JsonSchema)]
// #[schemars(example = "graphql_validation_example_1")]
pub struct GraphQLValidationPluginConfig {}
