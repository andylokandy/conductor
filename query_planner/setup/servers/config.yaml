sandbox:
  enabled: true
supergraph:
  introspection: true
homepage:
  enabled: false