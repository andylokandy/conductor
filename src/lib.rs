pub mod config;
pub mod endpoint;
pub mod gateway;
pub mod source;

use async_graphql::http::GraphiQLSource;
use axum::extract::State;
use axum::{Router, Server};
use config::load_config;
use endpoint::endpoint_runtime::EndpointRuntime;
use gateway::engine::Gateway;

use axum::http::Request;
use axum::response::{self, IntoResponse, Response};
use axum::routing::get;
use hyper::Body;

use tracing::debug;

pub async fn serve_graphiql_ide(req: Request<Body>) -> impl IntoResponse {
    response::Html(GraphiQLSource::build().endpoint(req.uri().path()).finish())
}

pub async fn handle_post(State(state): State<EndpointRuntime>, body: String) -> Response<Body> {
    let response = state.call(body).await;
    response.unwrap()
}

pub async fn run_services(config_file_path: String) {
    println!("gateway process started");
    println!("loading configuration from {}", config_file_path);
    let config_object = load_config(&config_file_path).await;
    println!("configuration loaded");

    tracing_subscriber::fmt()
        .with_max_level(config_object.logger.level.into_level())
        .init();

    debug!("loaded gateway config: {:?}", config_object);

    let gateway = Gateway::new(config_object);
    let mut http_router = Router::new();

    for (path, endpoint) in gateway.endpoints.into_iter() {
        let path_str = path.as_str();
        http_router = http_router.route(
            path_str,
            get(serve_graphiql_ide)
                .post(handle_post)
                .with_state(endpoint),
        )
    }

    println!("GraphiQL IDE: http://localhost:8000");

    Server::bind(&"127.0.0.1:8000".parse().unwrap())
        .serve(http_router.into_make_service())
        .await
        .unwrap();
}
