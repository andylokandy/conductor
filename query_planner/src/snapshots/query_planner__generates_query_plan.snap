---
source: src/lib.rs
expression: "(query_plan, response_vec)"
---
[
  {
    "steps": [
      {
        "Parallel": [
          {
            "SeqQueries": [
              {
                "service_name": "LOCATIONS",
                "query": "query AYOOO { locations { id name description isCool { really { yes __typename } __typename } __typename } }",
                "arguments": null
              },
              {
                "service_name": "REVIEWS",
                "query": "query AYOOO($representations: [_Any!]!) { \n  _entities(representations: $representations) {\n    ... on Location {\n      reviews { id comment rating __typename }\n    }\n   __typename } }",
                "arguments": null
              }
            ]
          },
          {
            "SeqQueries": [
              {
                "service_name": "LOCATIONS",
                "query": "query AYOOO { location(id: \"portugal\") { id photo __typename } }",
                "arguments": null
              }
            ]
          }
        ]
      }
    ]
  },
  [
    {
      "data": [
        {
          "locations": [
            {
              "__typename": "Location",
              "description": "The best country in the world",
              "id": "portugal",
              "isCool": {
                "__typename": "IsCool",
                "really": {
                  "__typename": "IsCoolReally",
                  "yes": true
                }
              },
              "name": "Portugal - European Union"
            },
            {
              "__typename": "Location",
              "description": "The worst country in the world",
              "id": "usa",
              "isCool": {
                "__typename": "IsCool",
                "really": {
                  "__typename": "IsCoolReally",
                  "yes": false
                }
              },
              "name": "USA - United States Of America"
            }
          ]
        },
        {
          "_entities": [
            {
              "__typename": "Location",
              "reviews": [
                {
                  "__typename": "Review",
                  "comment": "Wow! Portugal is amazing!",
                  "id": "1",
                  "rating": 9
                }
              ]
            },
            {
              "__typename": "Location",
              "reviews": [
                {
                  "__typename": "Review",
                  "comment": "USA could be better...",
                  "id": "2",
                  "rating": 6
                }
              ]
            }
          ]
        }
      ],
      "errors": null
    },
    {
      "data": [
        {
          "location": {
            "__typename": "Location",
            "id": "portugal",
            "photo": "https://...."
          }
        }
      ],
      "errors": null
    }
  ]
]
